pkg/util/log/line_buffer.go

```diff
@@ -0,0 +1,162 @@
+// syscall for every log line which is expensive if logging thousands of lines per second.
+type LineBufferedLogger struct {
+	buf     *threadsafeBuffer
+	entries atomic.Uint32
+	cap     uint32
+	w       io.Writer
+
+	onFlush func(entries uint32)
+}
+
+// Size returns the number of entries in the buffer.
+func (l *LineBufferedLogger) Size() uint32 {
+	return l.entries.Load()
+}
+
+// Write writes the given bytes to the line buffer, and increments the entries counter.
+// If the buffer is full (entries == cap), it will be flushed, and the entries counter reset.
+func (l *LineBufferedLogger) Write(p []byte) (n int, err error) {
+	// when we've filled the buffer, flush it
+	if l.Size() >= l.cap {
+		// Flush resets the size to 0
+		if err := l.Flush(); err != nil {
+			l.buf.Reset()
+			return 0, err
+		}
+	}
+
+	l.entries.Inc()
+
+	return l.buf.Write(p)
+}
+
+// Flush forces the buffer to be written to the underlying writer.
+func (l *LineBufferedLogger) Flush() error {
+	sz := l.Size()
+	if sz <= 0 {
+		return nil
+	}
+
+	// reset the counter
+	l.entries.Store(0)
+
+	// WriteTo() calls Reset() on the underlying buffer, so it's not needed here
+	_, err := l.buf.WriteTo(l.w)
+
+	// only call OnFlush callback if write was successful
+	if err == nil && l.onFlush != nil {
+		l.onFlush(sz)
+	}
+
+	return err
+}
+
+type LineBufferedLoggerOption func(*LineBufferedLogger)
+
+// WithFlushPeriod creates a new LineBufferedLoggerOption that sets the flush period for the LineBufferedLogger.
+func WithFlushPeriod(d time.Duration) LineBufferedLoggerOption {
+	return func(l *LineBufferedLogger) {
+		go func() {
+			tick := time.NewTicker(d)
+			defer tick.Stop()
+
+			for range tick.C {
+				l.Flush()
+			}
+		}()
+	}
+}
+
+// WithFlushCallback allows for a callback function to be executed when Flush() is called.
+// The length of the buffer at the time of the Flush() will be passed to the function.
+func WithFlushCallback(fn func(entries uint32)) LineBufferedLoggerOption {
+	return func(l *LineBufferedLogger) {
+		l.onFlush = fn
+	}
+}
+
+// WithPrellocatedBuffer preallocates a buffer to reduce GC cycles and slice resizing.
+func WithPrellocatedBuffer(size uint32) LineBufferedLoggerOption {
+	return func(l *LineBufferedLogger) {
+		l.buf = newThreadsafeBuffer(bytes.NewBuffer(make([]byte, 0, size)))
+	}
+}
+
+// NewLineBufferedLogger creates a new LineBufferedLogger with a configured capacity.
+// Lines are flushed when the context is done, the buffer is full, or the flush period is reached.
+func NewLineBufferedLogger(w io.Writer, cap uint32, opts ...LineBufferedLoggerOption) *LineBufferedLogger {
+	l := &LineBufferedLogger{
+		w:   w,
+		buf: newThreadsafeBuffer(bytes.NewBuffer([]byte{})),
+		cap: cap,
+	}
+
+	for _, opt := range opts {
+		opt(l)
+	}
+
+	return l
+}
+
+// threadsafeBuffer wraps the non-threadsafe bytes.Buffer.
+type threadsafeBuffer struct {
+	sync.RWMutex
+
+	buf *bytes.Buffer
+}
+
+// Read returns the contents of the buffer.
```

Comment:
The following should be more precise.
```suggestion
// Read reads up to len(p) bytes into p. It returns the number of bytes read (0 <= n <= len(p)) and any error encountered.
```